### PR TITLE
Feature: Form Templates

### LINKED ISSUE #1816: Redesign Create New Form modal steps
Due to some issues with applying the template to already created project we need to rethink current modal flow.

Design "Create New Form" modal steps:

* First: Choose source first:
  * "Start from Scratch" (current "Design in Form Builder")
  * "Use a template"
  * "Upload an XLSForm"
  * "Import an XLSForm via URL"
* Further (depending on source):
  * "Start from Scratch"
    * "Define Project"
    * open Form Builder
  * "Use a template"
    * template selector
    * "Define Project" (data filled up)
    * open Project Landing
  * "Upload an XLSForm"
    * file selector
    * "Define Project" (data filled up)
    * open Project Landing
  * "Import an XLSForm via URL"
    * url importer
    * "Define Project" (data filled up)
    * open Project Landing
    
During the step `"Define Project" (data filled up)"` the button could be ~"SKIP"/"APPROVE"~ "Continue" and when user edits some filled up inputs it changes to ~"SAVE & GO"~ "Save & Continue"?

### PR DESCRIPTION
**Pull request by magicznyleszek**
There will not be much (if any) changes to FE as for now, so I think @pmusaraj you can take a look :) Plus this merge goes to the `form-template` main branch, so further FE changes can be done there.

fixes #1816 #1810 #1634 #1633

### CODE REVIEW

**Comment by pmusaraj** [comment]
A few things I noticed from an initial overview on `unu`: 

- [ ] in library list, "Create project" should likely be outside of the ... menu for templates
- [ ] Instead of "Enter new name or leave blank to use ##templatename##" it should say "Enter the name of your new project (or leave blank to use ##templatename## as the project name)"
- [ ] after creating a project from a template, user should be taken to the newly-created project (same for the other way around, after creating template from a project, user should be taken to the library list)
- [ ] The notification "new survey has been created" should read "new project has been created" (let's use project consistently in our messaging)
- [ ] when in a project's form landing screen, under Replace with XLS, we should allow replacing the project from a template (it should be the third option in the modal, and "Replace with XLS" can be renamed "Replace project")
- [ ] in a project's form landing, please add "Create Template" as an option in the ... menu
- [ ] when creating/editing form templates, the metadata bar is not visible, and the settings cannot be edited (see https://github.com/kobotoolbox/kpi/issues/1815)
- [ ] when creating a template from a project and then creating a new project from that template, the "share sector/country" checkbox (which was checked in the initial project) isn't carried to the new project. (the other fields, description, sectory, country are carried over)

**Comment by magicznyleszek** [comment]
I've moved the "Create Project" option outside of the dropdown to the icons list next to it

**Comment by magicznyleszek** [comment]
Currently blocked by #1816

**Comment by magicznyleszek** [comment]
New TODO: 
* [ ] Form Builder nav https://app.atomic.io/d/82NlfhccJyO7

**Comment by pmusaraj** [comment]
@magicznyleszek 

I was getting an error in the library, when hitting the "Create Project" button next to a template: `Cannot read property 'cloneAsset' of undefined` in `utils.es6`. I'm not sure whether this ever worked, but I decided to move the `cloneAssetAsNewType` helper to mixins, since it's only ever used there. (`utils.es6` is more of a general-purpose, feature-agnostic file.) 

While doing that, I noticed that the clone dialogs use the word "form" everywhere, and I replaced that with the correct type of the asset (project, template, question, block), all four types are now listed in `constants.es6` for better translation management. 

I also made some small styling changes to the template list in the modal. 

Let me know when you're doing with the changes this requires to the form builder. My next step in testing this will be that, and making sure that template settings (description, sector, country) get transferred to the project.

**Comment by magicznyleszek** [comment]
@pmusaraj I made a change in `constants.es6` in 396a492 - please take a look :)

**Comment by magicznyleszek** [comment]
@magicznyleszek thing from standup:

* [ ] replace project should not overwrite name and settings

File: jsapp/js/components/formEditors.es6 [hunk 1]
```diff
@@ -37,330 +39,766 @@ import {
...
+      case this.STEPS.FORM_SOURCE: return t('Choose a source');
+      case this.STEPS.CHOOSE_TEMPLATE: return t('Choose template');
+      case this.STEPS.UPLOAD_FILE: return t('Upload XLSForm');
+      case this.STEPS.IMPORT_URL: return t('Import XLSForm');
+      case this.STEPS.PROJECT_DETAILS: return t('Project details');
+      default: return '';
+    }
+  }
+
+  getFilenameFromURI(url) {
```
Comment (jnm) [inline jsapp/js/components/formEditors.es6]: @magicznyleszek, it looks like this was introduced in https://github.com/kobotoolbox/kpi/commit/e32e706d0db1680ef3085fef93cf36c5d403df14 but isn't used anymore as of https://github.com/kobotoolbox/kpi/commit/33ffa8bea60efb4d0a723286403a7751a4ea06d4

File: jsapp/js/editorMixins/editableForm.es6 [hunk 2]
```diff
@@ -129,80 +138,52 @@ class FormSettingsBox extends React.Component {
...
 
-    if (this.state.editorState == 'existing') {
+    if (this.state.editorState === 'existing') {
       let uid = this.props.params.assetid;
       stores.allAssets.whenLoaded(uid, (asset) => {
-        let translations = (asset.content && asset.content.translations
-                            && asset.content.translations.slice(0)) || [];
+        this.setState({asset: asset});
+
+        // TODO investigate why the heck `slice(0)` is being used here,
```
[the remark this answers is not in the document]
Comment (magicznyleszek) [reply jsapp/js/editorMixins/editableForm.es6]: make sense, thanks :)

[the remark this answers is not in the document]
Comment (pmusaraj) [reply jsapp/js/editorMixins/editableForm.es6]: I'm not sure this is my code. In any case, I'm planning to move away from translations in the form builder in the 
https://github.com/kobotoolbox/kpi/tree/translations-U branch. I

File: jsapp/js/components/list.es6 [hunk 3]
```diff
@@ -275,8 +275,10 @@ class ListSearchSummary extends React.Component {
     this.setState(state);
   }
   render () {
-    var messages = [], modifier,
-        s = this.state;
+    var messages = [];
```
Comment (pmusaraj) [inline jsapp/js/components/list.es6]: It's not a big deal for this PR, but in general, let's avoid making these types of stylistic changes.

File: jsapp/js/components/formEditors.es6 [hunk 4]
```diff
@@ -761,6 +1159,11 @@ export class AddToLibrary extends React.Component {
       editorState: 'new',
       backRoute: '/library'
     };
+
+    if (this.props.location.pathname === '/library/new/template') {
```
Comment (pmusaraj) [inline jsapp/js/components/formEditors.es6]: @magicznyleszek can you try using the `isActiveRoute` method from the contextRouter mixin here?

Comment (magicznyleszek) [reply jsapp/js/components/formEditors.es6]: I did and it's not possible

Comment (pmusaraj) [reply jsapp/js/components/formEditors.es6]: ok.

Comment (jnm) [reply jsapp/js/components/formEditors.es6]: @magicznyleszek, there are people using KPI with a prefix, e.g. `https://my.server/kpi/library/new/template`; would this work with a setup like that?

Comment (pmusaraj) [reply jsapp/js/components/formEditors.es6]: It should be fine for KPI with prefixes, we've been using the same `this.props.location.pathname` construct in `formSubScreens.es6` for a while. I think this is an internal React location prop, independent of `window`.

**Comment by jnm** [comment]
Some minor issues to resolve before merging this

- [ ] @jnm to write backend changes to search filter for Whoosh workaround
    * https://github.com/kobotoolbox/kpi/pull/1813/files?w=1#diff-090a4263a189d134ef1ea609a77d3549R50
    * https://github.com/kobotoolbox/kpi/pull/1813/files?w=1#diff-164e86008eddfe3f43b3d1853925bedaR67
    * https://github.com/kobotoolbox/kpi/pull/1813/files?w=1#diff-164e86008eddfe3f43b3d1853925bedaR197
    * https://github.com/kobotoolbox/kpi/pull/1813/files?w=1#diff-06f4655d3effa64bf33171af20f51bbbR156
    * https://github.com/kobotoolbox/kpi/pull/1813/files?w=1#diff-024909a5199ee03202dd189ba9262f73R24
    * Some work already done:
        * https://github.com/kobotoolbox/kpi/pull/1813/files?w=1#diff-dfa79dbfb7f8b457062635c9bd6f80ecR132

- [ ] Typos in comments:
    * https://github.com/kobotoolbox/kpi/pull/1813/files?w=1#diff-90bd3d42a3830d82750a7bbea1468d8cR328
    * https://github.com/kobotoolbox/kpi/pull/1813/files?w=1#diff-1a6c3610e5a6ee4226ac3fda5d205aa3R3
- [ ] English doesn't really have this meaning of "aside" as a noun, making the name confusing
    * https://github.com/kobotoolbox/kpi/pull/1813/files?w=1#diff-1e9e28caf1b9bcc1aadb6854611c3b84R789
- [ ] Maybe replace "fueled" with "driven"?
    * https://github.com/kobotoolbox/kpi/pull/1813/files?w=1#diff-a812d6a1edf8c072009f19c32650acc8R192
    * https://github.com/kobotoolbox/kpi/pull/1813/files?w=1#diff-a812d6a1edf8c072009f19c32650acc8R233

**Pull request merged**